beacon-chain/p2p/gossipsub_peer_crawler.go

```diff
@@ -0,0 +1,515 @@
+		scorer:             scorer,
+	}
+	g.pingCh = make(chan enode.Node, 4*g.maxConcurrentPings)
+	g.pingSemaphore = semaphore.NewWeighted(int64(g.maxConcurrentPings))
+	g.crawledPeers = &crawledPeers{
+		g:        g,
+		byEnode:  make(map[enode.ID]*peerNode),
+		byPeerId: make(map[peer.ID]*peerNode),
+		byTopic:  make(map[gossipsubcrawler.Topic]map[peer.ID]struct{}),
+	}
+	return g, nil
+}
+
+func (g *GossipsubPeerCrawler) PeersForTopic(topic gossipsubcrawler.Topic) []*enode.Node {
+	g.crawledPeers.mu.RLock()
+	defer g.crawledPeers.mu.RUnlock()
+
+	peerIDs, ok := g.crawledPeers.byTopic[topic]
+	if !ok {
+		return nil
+	}
+
+	var peerNodes []*peerNode
+	seen := make(map[enode.ID]bool)
+	for peerID := range peerIDs {
+		peerNode, ok := g.crawledPeers.byPeerId[peerID]
+		if !ok {
+			continue
+		}
+		if peerNode.isPinged && g.peerFilter(peerNode.node) {
+			// Skip if we've already seen this enode ID
+			if seen[peerNode.id] {
+				continue
+			}
+			seen[peerNode.id] = true
+			peerNodes = append(peerNodes, peerNode)
+		}
+	}
+
+	// Sort peerNodes in descending order of their scores.
+	sort.Slice(peerNodes, func(i, j int) bool {
+		scoreI := g.scorer(peerNodes[i].peerID)
+		scoreJ := g.scorer(peerNodes[j].peerID)
+		return scoreI > scoreJ
+	})
+
+	nodes := make([]*enode.Node, len(peerNodes))
+	for i, pn := range peerNodes {
+		nodes[i] = pn.node
+	}
+
+	return nodes
+}
+
+func (g *GossipsubPeerCrawler) RemovePeerId(peerID peer.ID) {
+	g.crawledPeers.removePeerId(peerID)
+}
+
+func (g *GossipsubPeerCrawler) RemoveTopic(topic gossipsubcrawler.Topic) {
+	g.crawledPeers.removeTopic(topic)
+}
+
+// Start runs the crawler's loops in the background.
+func (g *GossipsubPeerCrawler) Start(te gossipsubcrawler.TopicExtractor) error {
+	if te == nil {
+		return errors.New("topic extractor is nil")
+	}
+	g.once.Do(func() {
+		g.topicExtractor = te
+		g.wg.Go(func() {
+			g.crawlLoop()
+		})
+		g.wg.Go(func() {
+			g.pingLoop()
+		})
+		g.wg.Go(func() {
+			g.cleanupLoop()
+		})
+	})
+
+	return nil
+}
+
+// Stop terminates the crawler.
+func (g *GossipsubPeerCrawler) Stop() {
+	g.cancel()
+	g.wg.Wait()
+}
+
+func (g *GossipsubPeerCrawler) pingLoop() {
+	for {
+		select {
+		case node := <-g.pingCh:
+			if err := g.pingSemaphore.Acquire(g.ctx, 1); err != nil {
+				return // Context cancelled, exit loop.
+			}
+			go func(node *enode.Node) {
+				defer g.pingSemaphore.Release(1)
+
+				if err := g.dv5.Ping(node); err != nil {
+					g.crawledPeers.removePeerOnPingFailure(node.ID())
+					return
+				}
+
+				g.crawledPeers.updateStatusToPinged(node.ID())
+			}(&node)
+
+		case <-g.ctx.Done():
+			return
+		}
+	}
+}
+
+func (g *GossipsubPeerCrawler) crawlLoop() {
+	ticker := time.NewTicker(g.crawlInterval)
+	defer ticker.Stop()
+
+	g.crawl()
+	for {
+		select {
+		case <-ticker.C:
+			g.crawl()
+		case <-g.ctx.Done():
+			return
+		}
+	}
+}
+
+func (g *GossipsubPeerCrawler) crawl() {
+	ctx, cancel := context.WithTimeout(g.ctx, g.crawlTimeout)
```

Comment:
What happens if `g.crawlTimeout >>> g.crawlInterval`?

Reply:
So
```go
//     - crawlInterval: How often to run the discovery crawl
```
is not true anymore?

It represents now the duration between end of the previous crawl and the start of the next one?